### PR TITLE
Update boundary_project_solution to project on edge and node boundary IDs

### PR DESCRIPTION
**Pull request by dknez**
Follow the approach from DofMap::apply_dirichlet_impl to make boundary_project_solution work with edge and node boundary IDs.

### CODE REVIEW

**Comment by dknez** [comment]
@roystgnr is this PR is OK with you? It's based on some changes we made earlier to apply_dirichlet_impl.

**Comment by dknez** [comment]
I agree that's it's a good idea to add a unit test. I've now added a unit test that covers boundary projection on sides, edges, and nodes.

File: tests/systems/systems_test.C [hunk 1]
```diff
@@ -214,6 +216,145 @@ public:
...
+    boundary_ids.insert(SIDE_BOUNDARY_ID);
+    std::vector<unsigned int> variables;
+    variables.push_back(u_var);
+    ZeroFunction<> zf;
+    sys.boundary_project_solution(boundary_ids, variables, &zf);
+
+    // We set the solution to be 1 everywhere and then zero on specific boundary
+    // nodes, so the final l1 norm of the solution is the difference between the
+    // number of nodes in the mesh and the number of nodes we zero on.
+    Real ref_l1_norm = static_cast<Real>(mesh.n_nodes()) - static_cast<Real>(projected_nodes_set.size());
```
Comment (roystgnr) [inline tests/systems/systems_test.C]: That's clever.

**Comment by roystgnr** [comment]
Nobody ever expects their test cases to run on a --enable-parmesh build.  :laughing:

The failure I get looks like a one-line fix, though; I'll put in a PR shortly.

**Pull request merged**